cnvlib/ngfrills/faidx.py

```diff
@@ -2,107 +2,21 @@
 from __future__ import absolute_import, division, print_function
 
 import logging
-import math
-import os
 from itertools import groupby
-
-import pysam
-from Bio._py3k import map
-
-from .shared import is_newer_than
-
-
-def ensure_fasta_index(fasta_fname):
-    """Ensure a FASTA file is indexed for samtools, to enable fast lookup."""
-    fai_fname = fasta_fname + '.fai'
-    if not is_newer_than(fai_fname, fasta_fname):
-        logging.info("Indexing FASTA file %sw", fasta_fname)
-        pysam.faidx(fasta_fname)
-    assert os.path.isfile(fai_fname), "Failed to generate index " + fai_fname
-    return fai_fname
-
-
-def read_fasta_index(fasta_fname):
-    """Load a FASTA file's index.
-
-    Returns a dict of:
-        {seq_id: (length, offset, chars_per_line, bytes_per_line), ...}
-
-    The index file contains, in one row per sequence, tab-separated columns:
-
-        - sequence identifier
-        - length
-        - offset of the first sequence character in the file
-        - number of characters per line
-        - number of bytes per line (including the end-of-line character)
-
-    With this information, we can easily compute the byte offset of the i-th
-    character of a sequence in a file by looking at its index record. We skip to
-    this byte offset in the file and from there, we can read the necessary
-    sequence characters.
-
-    See: http://trac.seqan.de/wiki/Tutorial/IndexedFastaIO:
-    """
-    index = {}
-    fai_fname = ensure_fasta_index(fasta_fname)
-    with open(fai_fname) as faifile:
-        for line in faifile:
-            fields = line.rstrip().split('\t')
-            seq_id = fields[0]
-            assert seq_id not in index, "Duplicate ID: " + seq_id
-            index[fields[0]] = tuple(map(int, fields[1:]))
-    return index
+from pyfaidx import Fasta
 
 
 def fasta_extract_regions(fa_fname, intervals):
     """Extract an iterable of regions from an indexed FASTA file.
 
-    Input: indexed FASTA file name; iterable of (seq_id, start, end) (1-based)
+    Input: FASTA file name; iterable of (seq_id, start, end) (1-based)
     Output: iterable of string sequences.
     """
-    index = read_fasta_index(fa_fname)
-    with open(fa_fname, 'rb') as fa_file:
+    with Fasta(fa_fname, as_raw=True) as fa_file:
         for chrom, rows in groupby(intervals, lambda cse: cse[0]):
-            # Seek to chrom offset in FASTA
-            try:
-                seq_len, offset, chars_per_line, bytes_per_line = index[chrom]
-            except KeyError:
-                raise ValueError("Sequence ID '" + chrom + "' is not in FASTA "
-                                 + "file " + fa_fname)
             logging.info("Extracting sequences from chromosome %s", chrom)
-            eol_size = bytes_per_line - chars_per_line  # Handle \n\r, \n
             for _chrom, start, end in rows:
```

Comment:
They are 0-based internally. I only convert to 1-based when writing to another file format.